### PR TITLE
drivers: gpio: axp192: GPIO driver implementation

### PR DESCRIPTION
**Pull request by MrMarteng**
AXP192 is a small power management IC, that also features 5 GPIOS.
Besides GPIO driver this commit also includes needed modifications in axp192 regulator and mfd driver as LDOIO0 functioanlity is multiplexed with GPIO0 pin.

Signed-off-by: Martin Kiepfer <mrmarteng@teleschirm.org>

### CODE REVIEW

File: drivers/gpio/gpio_axp192.c [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+{
+	int ret;
+	uint32_t value;
+
+	ret = gpio_axp192_port_get_raw(dev, &value);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return gpio_axp192_port_set_masked_raw(dev, pins, ~value);
```
Comment (henrikbrixandersen) [inline drivers/gpio/gpio_axp192.c:140]: Does this require locking?

File: drivers/mfd/mfd_axp192.c [hunk 2]
```diff
@@ -50,12 +153,455 @@ static int mfd_axp192_init(const struct device *dev)
...
+	}
+
+	ret = i2c_reg_read_byte_dt(&(config->i2c), gpio_reg_desc[gpio].reg, &reg_fnc);
+	if (ret != 0) {
+		return ret;
+	}
+
+	switch (gpio) {
+	case 0U:
+	case 1U:
```
Comment (henrikbrixandersen) [inline drivers/mfd/mfd_axp192.c:174]: You need to use `__fallthrough` for all fall-through cases.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Looks good! A couple of minor comments/questions:

File: drivers/mfd/mfd_axp192.c (hunk 2 above)
Comment (MrMarteng) [reply drivers/mfd/mfd_axp192.c]: Fixed

File: drivers/gpio/gpio_axp192.c (hunk 1 above)
Comment (MrMarteng) [reply drivers/gpio/gpio_axp192.c]: Very good catch. Your are absolutely right, there was missing protection for this rmw-action.
I also reviewed the mfd driver of possible issues, but couldn't find any further segments that also need protection.
Thank you for that valuable feedback.

**Comment by MrMarteng** [comment]
Any further comment?

**Comment by MrMarteng** [comment]
Hey guys, this is a kind reminder to support me on this PR.

**Comment by gmarull** [comment]
> Hey guys, this is a kind reminder to support me on this PR.

lgtm, just some minor comments. Believe it or not, GPIO area does not have a maintainer right now...

File: drivers/gpio/gpio_axp192.c [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+					gpio_flags_t flags)
+{
+	const struct gpio_axp192_config *config = dev->config;
+	int ret;
+	enum axp192_gpio_func func;
+
+	if (pin >= AXP192_GPIO_NUM_PINS) {
+		LOG_ERR("Invalid gpio pin (%d)", pin);
+		return -EINVAL;
+	}
```
[the remark this answers is not in the document]
Comment (MrMarteng) [reply drivers/gpio/gpio_axp192.c:79]: The number is physically restricted and this information is also needed on mfd layer. I can use the `ngpios` property in gpio driver but would it really be the better solution to have 2 definitions then?

File: drivers/mfd/mfd_axp192.c [hunk 4]
```diff
@@ -50,12 +153,495 @@ static int mfd_axp192_init(const struct device *dev)
 	return 0;
 }
 
+int mfd_axp192_gpio_func_get(const struct device *dev, uint8_t gpio, enum axp192_gpio_func *func)
```
Comment (gmarull) [inline drivers/mfd/mfd_axp192.c:156]: question: why is this part of the mfd layer? who else is using this?

File: drivers/mfd/mfd_axp192.c [hunk 5]
```diff
@@ -50,12 +153,495 @@ static int mfd_axp192_init(const struct device *dev)
...
+	uint8_t reg_fnc;
+
+	if (gpio >= AXP192_GPIO_NUM_PINS) {
+		LOG_ERR("Invalid gpio (%d)", gpio);
+		return -EINVAL;
+	}
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
```
Comment (gmarull) [inline drivers/mfd/mfd_axp192.c:169]: this is translating GPIO API requirements to the mfd layer... wrong

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/mfd/mfd_axp192.c (hunk 4 above)
Comment (MrMarteng) [reply drivers/mfd/mfd_axp192.c:156]: It is used by regulator and gpio drivers as some pins of the pmu are multiplexed. And it also was a recommendation in my regulator PR (https://github.com/zephyrproject-rtos/zephyr/pull/58809#discussion_r1217764673) .

File: drivers/mfd/mfd_axp192.c (hunk 5 above)
Comment (MrMarteng) [reply drivers/mfd/mfd_axp192.c:169]: I See. Thank you.

File: drivers/gpio/gpio_axp192.c [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+{
+	return gpio_axp192_port_set_masked_raw(dev, pins, pins);
+}
+
+static int gpio_axp192_port_clear_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	return gpio_axp192_port_set_masked_raw(dev, pins, 0);
+}
+
+static inline int gpio_axp192_configure(const struct device *dev, gpio_pin_t pin,
```
Comment (gmarull) [inline drivers/gpio/gpio_axp192.c:78]:
```suggestion
static int gpio_axp192_configure(const struct device *dev, gpio_pin_t pin,
```

File: tests/drivers/build_all/gpio/prj.conf [hunk 7]
```diff
@@ -4,4 +4,5 @@ CONFIG_TEST_USERSPACE=y
 CONFIG_I2C=y
 CONFIG_GPIO_PCA95XX_INTERRUPT=y
 CONFIG_SPI=y
+CONFIG_MFD=y
```
Comment (gmarull) [inline tests/drivers/build_all/gpio/prj.conf:7]: redundant, driver must select MFD

**Review by gmarull — COMMENTED** [review]
changes to regulator driver should go into its own commit maybe?

File: drivers/mfd/mfd_axp192.c [hunk 8]
```diff
@@ -19,10 +21,111 @@ LOG_MODULE_REGISTER(mfd_axp192, CONFIG_MFD_LOG_LEVEL);
...
+/* Pull-Down enable parameters */
+#define AXP192_GPIO0_PULLDOWN_ENABLE 0x01U
+#define AXP192_GPIO1_PULLDOWN_ENABLE 0x02U
+#define AXP192_GPIO2_PULLDOWN_ENABLE 0x04U
+
+/* GPIO Value parameters */
+#define AXP192_GPIO0_INPUT_VAL      0x10U
+#define AXP192_GPIO1_INPUT_VAL      0x20U
+#define AXP192_GPIO2_INPUT_VAL      0x40U
+#define AXP192_GPIO012_INTPUT_SHIFT 4U
```
Comment (aasinclair) [inline drivers/mfd/mfd_axp192.c:70]: nit: Should this be INPUT?

File: drivers/mfd/mfd_axp192.c [hunk 9]
```diff
@@ -50,12 +153,463 @@ static int mfd_axp192_init(const struct device *dev)
...
+	ret = mfd_axp192_gpio_read_port(dev, &port_val);
+	if (ret != 0) {
+		return ret;
+	}
+
+	if ((port_val & (1u << pin)) != 0) {
+		*value = true;
+	} else {
+		*value = false;
+	}
```
Comment (aasinclair) [inline drivers/mfd/mfd_axp192.c:532]:
```suggestion
	*value = (port_val & BIT(pin)) != 0;
```

File: drivers/gpio/gpio_axp192.c [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+		ret = mfd_axp192_gpio_func_ctrl(config->mfd, dev, pin, func);
+		if (ret != 0) {
+			return ret;
+		}
+
+		/* Set init value */
+		if ((flags & GPIO_OUTPUT_INIT_LOW) != 0) {
+			ret = mfd_axp192_gpio_write_pin(config->mfd, pin, false);
+		} else if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0) {
+			ret = mfd_axp192_gpio_write_pin(config->mfd, pin, true);
```
Comment (aasinclair) [inline drivers/gpio/gpio_axp192.c:112]: Is this the only place you are using mfd_axp192_gpio_write_pin?

If so you could replace it with a call to `mfd_axp192_gpio_write_port(config->mfd, BIT(pin),...)` and remove mfd_axp192_gpio_write_pin entirely.

File: drivers/mfd/mfd_axp192.c [hunk 11]
```diff
@@ -50,12 +153,456 @@ static int mfd_axp192_init(const struct device *dev)
...
+	gpio_output_val = (gpio012_val & AXP192_GPIO012_OUTPUT_MASK);
+	gpio_output_val |= ((gpio34_val & AXP192_GPIO34_OUTPUT_MASK) << 3u);
+
+	*value = gpio_input_val & ~(data->gpio_mask_output);
+	*value |= (gpio_output_val & data->gpio_mask_output);
+
+	return 0;
+}
+
+int mfd_axp192_gpio_read_pin(const struct device *dev, uint8_t pin, bool *value)
```
Comment (aasinclair) [inline drivers/mfd/mfd_axp192.c:513]: Is this function needed?

**Review by aasinclair — COMMENTED** [review]
This is looking really good

**Review by aasinclair — APPROVED** [review]

**Comment by MrMarteng** [comment]
@gmarull @henrikbrixandersen would you like to approve the PR as well?

**Review by gmarull — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Pull request merged**